File: stacks/pox.js
import axios from 'axios';
import { 
  deserializeCV
} from '@stacks/transactions';
import { c32address } from 'c32check'; // Import the c32address function from c32check
import { createObjectCsvWriter } from 'csv-writer';
import fs from 'fs';

// API base URL for Stacks mainnet
const STACKS_API_URL = 'https://api.hiro.so';
const contractAddress = "SP000000000000000000002Q6VF78";
const contractName = "pox-4";
const sender = "SP3TRVBX53CN78AS8C3HNTM3GPNDHGA34F9M7MAH2";

// CSV Writer setup
const csvWriter = createObjectCsvWriter({
  path: './stackers.csv',
  header: [
    { id: 'index', title: 'Index' },
    { id: 'poxAddrHash', title: 'PoX Address Hash' },
    { id: 'poxAddrVersion', title: 'PoX Address Version' },
    { id: 'signer', title: 'Signer' },
    { id: 'ustx', title: 'Total uSTX' },
    { id: 'stackerAddress', title: 'Stacker Address' }
  ]
});

// Parse Clarity value from a hex string
function parseClarityValue(hexStr) {
  if (hexStr.startsWith('0x')) {
    hexStr = hexStr.slice(2);
  }

  const buffer = Buffer.from(hexStr, 'hex');
  return deserializeCV(buffer);
}

// Convert value to Clarity Hex
function generateHex(value) {
  const hexValue = value.toString(16);
  const paddedValue = hexValue.padStart(33, '0');
  const finalHex = '01' + paddedValue.slice(1);
  return '0x' + finalHex;
}

// Decode Clarity Hex to integer
function decodeHex(value) {
  if (value.startsWith('0x')) {
    value = value.slice(2);
  }

  if (value.startsWith('01')) {
    value = value.slice(2);
  }

  const strippedValue = value.replace(/^0+/, '');
  if (!strippedValue) return 0;

  return parseInt(strippedValue, 16);
}

// Function to get the number of stackers for a specific cycle
async function getNoStackers(value) {
  const functionName = 'get-reward-set-size';
  const url = `${STACKS_API_URL}/v2/contracts/call-read/${contractAddress}/${contractName}/${functionName}`;

  const payload = {
    sender: sender,
    arguments: [generateHex(value)]
  };

  try {
    const response = await axios.post(url, payload);
    console.log(`Status Code: ${response.status}`);

    const result = response.data.result;
    const decodedResult = decodeHex(result);

    if (decodedResult) {
      console.log(`Number of stackers: ${decodedResult}`);
      return decodedResult;
    } else {
      console.log("No result found in response.");
      return null;
    }
  } catch (error) {
    console.error(`Error retrieving number of stackers: ${error}`);
  }
}

// Function to get stacker info by cycle and index
async function getStackersByCycle(cycle, index) {
  const functionName = 'get-reward-set-pox-address';
  const url = `${STACKS_API_URL}/v2/contracts/call-read/${contractAddress}/${contractName}/${functionName}`;

  const payload = {
    sender: sender,
    arguments: [generateHex(cycle), generateHex(index)]
  };

  try {
    const response = await axios.post(url, payload);

    const result = response.data.result;

    if (result) {
      const clarityValue = parseClarityValue(result);
      const clarityData = clarityValue.value.data;

      // Decoding pox-addr
      const poxAddr = clarityData['pox-addr'].data;
      const hashbytesHex = poxAddr.hashbytes.buffer.toString('hex');
      const versionHex = poxAddr.version.buffer.toString('hex');

      // Decoding signer
      const signer = clarityData['signer'];
      const signerHex = signer.buffer ? signer.buffer.toString('hex') : 'Signer is missing';

      // Decoding stacker
      const stacker = clarityData['stacker'];
      let stackerAddress = null;

      if (stacker.type === 10 && stacker.value.type === 5) {
        const address = stacker.value.address;
        const stackerVersion = address.version;
        const stackerHash160 = address.hash160;
        stackerAddress = c32address(stackerVersion, stackerHash160);
      }

      // Ensure Total uSTX is a number, not an object
      const totalUstx = clarityData['total-ustx'].value.toString(); // Convert to string to avoid [object Object]

      return {
        index: index,
        poxAddrHash: `0x${hashbytesHex}`,
        poxAddrVersion: `0x${versionHex}`,
        signer: `0x${signerHex}`,
        ustx: totalUstx,
        stackerAddress: stackerAddress
      };

    } else {
      console.log(`No result found for cycle ${cycle}, index ${index}.`);
      return null;
    }
  } catch (error) {
    console.error(`Error parsing response for cycle ${cycle}, index ${index}: ${error.message}`);
    return null;
  }
}

// Function to get total uSTX stacked for a cycle
async function getCycleData(cycle) {
  const functionName = 'get-total-ustx-stacked';
  const url = `${STACKS_API_URL}/v2/contracts/call-read/${contractAddress}/${contractName}/${functionName}`;

  const payload = {
    sender: sender,
    arguments: [generateHex(cycle)]
  };

  try {
    const response = await axios.post(url, payload);
    console.log(`Status Code: ${response.status}`);
    console.log("Response Text:");
    console.log(response.data);

    const result = response.data.result;
    const decodedResult = decodeHex(result) / 1e6;

    if (decodedResult) {
      console.log(`Total uSTX stacked: ${decodedResult}`);
    } else {
      console.log("No result found in response.");
    }
  } catch (error) {
    console.error(`Error retrieving cycle data: ${error}`);
  }
}

// Main execution
(async function () {
  const cycle = 94;

  // Get total uSTX stacked for the cycle
  await getCycleData(cycle);

  // Get the number of stackers for the cycle
  const numberOfStackers = await getNoStackers(cycle);

  if (numberOfStackers !== null) {
    const stackersData = [];

    // Collect stacker data for each index
    for (let index = 0; index < numberOfStackers; index++) {
      const stackerInfo = await getStackersByCycle(cycle, index);
      if (stackerInfo) {
        stackersData.push(stackerInfo);
      }
    }

    // Write the data to CSV
    if (stackersData.length > 0) {
      await csvWriter.writeRecords(stackersData);
      console.log('CSV file successfully written.');
    } else {
      console.log('No stacker data found to write to CSV.');
    }
  }
})();

File: python/pox.py
import requests

# API base URL for Stacks mainnet
STACKS_API_URL = 'https://api.hiro.so'
contract_address = "SP000000000000000000002Q6VF78"
contract_name = "pox-4"
sender = "SP3TRVBX53CN78AS8C3HNTM3GPNDHGA34F9M7MAH2" 

# Function to get the number of stackers for a specific cycle
def get_no_stackers(value):
    function_name = "get-reward-set-size"

    # Create the full API endpoint URL
    url = f"{STACKS_API_URL}/v2/contracts/call-read/{contract_address}/{contract_name}/{function_name}"

    # Payload for the specified cycle
    payload = {
        "sender": sender,  # Sender address
        "arguments": [generate_hex(value)]  # Hex argument for the cycle
    }

    # Make the POST request
    response = requests.post(url, json=payload)

    # Print the status code and response for debugging
    print(f"Status Code: {response.status_code}")
    print("Response Text:")
    print(response.text)  # Print raw response

    # If the response is successful, attempt to parse the result
    if response.status_code == 200:
        try:
            response_json = response.json()
            result = response_json.get('result', None)
            result = decode_hex(result)
            
            if result is not None:
                # Print and return the number of stackers
                print(f"Number of stackers: {result}")
                return result
            else:
                print("No result found in response.")
                return None
        except Exception as e:
            print(f"Error parsing response: {e}")
            return None
    else:
        print("Failed to retrieve data. Please check the API request.")
        return None

# Function to get stacker information for a specific cycle and index
def get_stackers_by_cycle(cycle, index):

    function_name = "get-reward-set-pox-address"

    # Create the full API endpoint URL
    url = f"{STACKS_API_URL}/v2/contracts/call-read/{contract_address}/{contract_name}/{function_name}"

    # Payload for the cycle and index
    payload = {
        "sender": sender,  
        "arguments": [generate_hex(cycle), generate_hex(index)]  # Cycle and index arguments in hex
    }

    # Make the POST request
    response = requests.post(url, json=payload)

    # Print the status code and response for debugging
    print(f"Status Code: {response.status_code}")
    print(f"Cycle {cycle}, Index {index} Response Text:")
    print(response.text)  # Print raw response

    # If the response is successful, attempt to parse the result
    if response.status_code == 200:
        try:
            response_json = response.json()
            result = response_json.get('result', None)
            
            if result:
                # Print the stacker info
                print(f"Stacker info (cycle {cycle}, index {index}): {result}")
            else:
                print(f"No result found for cycle {cycle}, index {index}.")
        except Exception as e:
            print(f"Error parsing response for cycle {cycle}, index {index}: {e}")
    else:
        print(f"Failed to retrieve data for cycle {cycle}, index {index}. Please check the API request.")

def get_cycle_data(cycle):
    function_name = "get-total-ustx-stacked"

    # Create the full API endpoint URL
    url = f"{STACKS_API_URL}/v2/contracts/call-read/{contract_address}/{contract_name}/{function_name}"

    # Payload for the specified cycle
    payload = {
        "sender": sender,  # Sender address
        "arguments": [generate_hex(cycle)]  # Hex argument for the cycle
    }

    # Make the POST request
    response = requests.post(url, json=payload)

    # Print the status code and response for debugging
    print(f"Status Code: {response.status_code}")
    print("Response Text:")
    print(response.text)  # Print raw response

    # If the response is successful, attempt to parse the result
    if response.status_code == 200:
        try:
            response_json = response.json()
            result = response_json.get('result', None)
            result = decode_hex(result) / 1e6  # Convert to appropriate units
            
            if result:
                # Print the total uSTX stacked
                print(f"Total uSTX stacked: {result}")
            else:
                print("No result found in response.")
        except Exception as e:
            print(f"Error parsing response: {e}")
    else:
        print("Failed to retrieve data. Please check the API request.")

def generate_hex(value):
    # Convert the value to hexadecimal and remove the "0x" prefix
    hex_value = format(value, 'x')
    
    # Pad with leading zeroes to ensure it's 33 characters
    padded_value = hex_value.zfill(33)
    
    # Prepend "01" in front and adjust the length accordingly
    final_hex = "01" + padded_value[1:]  # Replace the first zero with '01'
    
    # Add the "0x" prefix
    return "0x" + final_hex

def decode_hex(value):
    # Ensure the input starts with '0x' and remove it
    if value.startswith('0x'):
        value = value[2:]
    
    # Remove leading '01' (first two characters) and all leading zeros afterward
    if value.startswith('01'):
        value = value[2:]  # Remove '01'
    
    # Strip any remaining leading zeros
    stripped_value = value.lstrip('0')
    
    if stripped_value == '':
        return 0
    
    # Convert the remaining hex string to decimal
    decimal_value = int(stripped_value, 16)
    
    return decimal_value

# Main execution
if __name__ == "__main__":
    cycle = 94

    # Get total uSTX stacked for the cycle
    get_cycle_data(cycle)
    
    # Get the number of stackers for the cycle
    number_of_stackers = get_no_stackers(cycle)
    
    # If the number of stackers is retrieved successfully, iterate and get each stacker's info
    if number_of_stackers is not None:
        for index in range(number_of_stackers):
            get_stackers_by_cycle(cycle, index)
